topcoffea/modules/dataDrivenEstimation.py

```diff
@@ -0,0 +1,161 @@
+                        scale_dict[(appl, sample)] = 1000*get_lumi('20'+year)/smweight
+                histo.scale( scale_dict, axis=('appl','sample'))
+
+
+                # now for each year we actually perform the subtraction and integrate out the application regions
+                newhist=None
+                for ident in histo.identifiers('appl'):
+                    hAR=histo.integrate('appl', ident)
+
+                    if 'isAR' not in ident.name:
+                        # if we are in the signal region, we just take the 
+                        # whole histogram integrating out the application region axis
+                        if newhist==None:
+                            newhist=hAR
+                        else:
+                            newhist=newhist+hAR
+                    else:
+                        # if we are in the application region, we also integrate the application region axis
+                        # and construct the new sample 'nonprompt'
+
+                        # we look at data only, and rename it to fakes
+                        newNameDictData=defaultdict(list); newNameDictNoData=defaultdict(list)
+                        addedNonPrompts=[]
+                        for sample in hAR.identifiers('sample'):
+                            match = pattern.search(sample.name)
+                            sampleName=match.group('sample')
+                            year=match.group('year')
+                            nonPromptName='nonpromptUL%s'%year
+                            if self.dataName==sampleName:
+                                newNameDictData[nonPromptName].append(sample.name)
+                                addedNonPrompts.append( (nonPromptName, year)) # keep it to rescale later
+                            else: 
+                                # To do: hard code a list of samples we actually want to subtract? 
+                                # if we run over, say, ttbar mc it doesn't make sense to subtract that
+                                newNameDictNoData[nonPromptName].append(sample.name)
+                        
+                        hFakes=hAR.group('sample',  hist.Cat('sample','sample'), newNameDictData)
+                    
+                        # now we take all the stuff that is not data in the AR to make the prompt subtraction and assign them to nonprompt.
+                        hPromptSub=hAR.group('sample', hist.Cat('sample','sample'), newNameDictNoData )
+                        
+                        # now we actually make the subtraction
+                        hPromptSub.scale(-1)
+                        hFakes=hFakes+hPromptSub
+
+                        #scale back by 1/lumi as if it were a MC, so it can be used transparently downstream
+                        scaleDict={}
+                        for name, year in addedNonPrompts:
+                            scaleDict[name]=1/(1000*get_lumi(year))
```

Comment:
@sscruz, the code is crashing here for me, I think because `year` is `17` while the key in the lumi json is `2017`.